Support non-destructive push for custom data fields

mjriley opened the pull request:
## Technical Summary
This is the second in a set of PRs to support ERM Non Destructive Push.  Associated ticket found [here](https://dimagi-dev.atlassian.net/browse/SAAS-13622).  This PR contains code to handle ERM for Custom Data Fields.

## Safety Assurance

### Safety story
These changes have been live on staging for months now. I've done local testing and the changes are backed by automated tests. Custom Data Fields was a model that previously did not track whether data was synced or not, so the new migrations can be rolled back without impacting existing code.

### Automated test coverage

Several new test suites created:
_custom_data_fields/tests/test_edit_model_
_custom_data_fields/tests/test_models_
_linked_domain/tests/test_update_custom_data_fields_

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-5375


### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

biyeun commented on corehq/apps/custom_data_fields/static/custom_data_fields/js/custom_data_fields.js:
```diff
@@ -127,7 +158,7 @@ hqDefine('custom_data_fields/js/custom_data_fields', [
 
     function CustomDataFieldsModel(options) {
-        assertProperties.assertRequired(options, ['custom_fields', 'custom_fields_profiles']);
+        assertProperties.assertRequired(options, [ 'custom_fields', 'custom_fields_profiles']);
```
nit extra whitespace character

millerdev commented on corehq/apps/linked_domain/tests/test_update_custom_data_fields.py:
```diff
@@ -85,6 +90,189 @@ def test_update(self):
+
+        fields = self._get_fields(UserFieldsView.field_type)
+        expected_field = Field(slug='a', label='label2', is_synced=True)
+        self.assertEqual(len(fields), 1)
+        self.assertEqual(fields[0].to_dict(), expected_field.to_dict())
```
nit: I often find failure output to be more informative when it shows more than a simple length mismatch. Consider this alternative:
```suggestion
        self.assertEqual([f.to_dict() for f in fields], [expected_field.to_dict()])
```
The same applies to other tests as well.

mjriley replied:
It may be just how I interpret things, but I think reframing it this way obscures the assertion. For me, the two assertions state: check that there is exactly 1 field, and that it matches what is specified. When I see the list comprehension, despite knowing what the test is supposed to do, I think I start to question why we're iterating over all fields when I only want to look at one. 

I think the other thing is that I don't necessarily expect failing tests to give the data necessary to correct the test; rather, they should provide enough information to understand what is going wrong and start investigating. I think seeing that the test expected 1 record, but received 2, is straight-forward enough to begin investigating.

millerdev commented on corehq/apps/linked_domain/tests/test_update_custom_data_fields.py:
```diff
@@ -85,6 +90,189 @@ def test_update(self):
+        # Due to this, I'm just faking out 'has_users_assigned' to always believe that the profile is in use.
+
+        # create a profile
+        fields = [Field(slug='a', is_synced=True), ]
```
nit
```suggestion
        fields = [Field(slug='a', is_synced=True)]
```

mjriley replied:
appears to be outdated -- `is_synced` was replaced with `upstream_id`. I'll try to do better about removing historical artifacts in the future.

mjriley commented on corehq/apps/custom_data_fields/static/custom_data_fields/js/custom_data_fields.js:
```diff
@@ -39,6 +40,9 @@ hqDefine('custom_data_fields/js/custom_data_fields', [
         self.regex = ko.observable(options.regex);
         self.regex_msg = ko.observable(options.regex_msg);
+        self.isSynced = ko.observable(options.is_synced);
+        self.isEditable = ko.observable(!options.is_synced);
```
I think I left a bit of an unclean history here? This seems like it was a previous commit, and the current code got rid of `isSynced` and made `isEditable` a `pureComputed`.

AmitPhulera committed:
pull out index consts to const file

AmitPhulera committed:
import and use the consts from const file

AmitPhulera committed:
update check

kaapstorm committed:
Don't overwrite gettext_lazy with an unnamed var

kaapstorm committed:
isort, drop unused imports

AmitPhulera committed:
add support to delete index after reindex is complete

kaapstorm committed:
Merge pull request #33166 from dimagi/nh/small_fix

Small fix

minhaminha committed:
Merge pull request #33161 from dimagi/ml/mac-setup-doc-updates

Update Mac Dev Setup doc

Jtang-1 committed:
bug: matches xpath that have escaped chracters when popping sort fields

MartinRiese committed:
Use observable to change class

MartinRiese committed:
Use marionette to change class

AmitPhulera committed:
Merge pull request #33146 from dimagi/ap/es/remove-size-0-from-aggs

Ensure TermsAggregation size is never 0

AmitPhulera committed:
add support for batch size in reindex

AmitPhulera committed:
pass in batch size from elastic_sync_multiplexed

AmitPhulera committed:
update docstrings of the command

AmitPhulera committed:
ask for index cname before deleting index

AmitPhulera committed:
use exponential timeoff while waiting for index to turn green

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

MartinRiese committed:
Merge pull request #33164 from dimagi/riese/unscrolling_map

Change the position of the case list map to sticky

MartinRiese committed:
WebApp -> Web Apps

Co-authored-by: Jenny Schweers <orange.jenny@gmail.com>

AmitPhulera committed:
add support to get index size on disk and required esimates for reindex

AmitPhulera committed:
expose the functionality to command elastic_sync_multiplexed

MartinRiese committed:
Address review comments

* Remove style form hidden fields
* Better variable names
* template_width is not needed when inheriting from HideShortColumn

Jtang-1 committed:
refactor: use _groupby

MartinRiese committed:
Fix test

MartinRiese committed:
Fix test

AmitPhulera committed:
Apply suggestions from code review

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera committed:
remove sleep_time from _wait_for_index_to_get_healthy

AmitPhulera committed:
lint

MartinRiese committed:
Add validator for multiple uses of Address format

orangejenny committed:
Removed case tiles ID mapping validation

orangejenny committed:
Fixed lint error

orangejenny committed:
Removed unused var

MartinRiese committed:
Fix mistake in renaming

Jtang-1 committed:
bug: creates indexed mapping instead of depending on offsetting index

Jtang-1 committed:
bug: maps index instead of relying on order
not all tiles are displayed so indices defined inheaderRowIndices are not necessarily
a consecutive sequence

orangejenny committed:
Merge pull request #33170 from dimagi/orangejenny-patch-11

Removed case tiles ID mapping validation

orangejenny committed:
Merge pull request #33155 from dimagi/jls/render-html-entities-in-tiles

Render line breaks properly in Web Apps markdown

orangejenny committed:
Merge branch 'master' into jls/web-apps-image-enum

Jtang-1 committed:
Merge branch 'master' into em/case-search-tiles-action-fix

MartinRiese committed:
Merge pull request #33149 from dimagi/riese/detail_popup

Add address_popup format

orangejenny committed:
Merge branch 'master' into jls/case-tile-template-preview

MartinRiese committed:
Merge branch 'master' into riese/validate_address_popup_unique

orangejenny committed:
Merge pull request #33096 from dimagi/jls/web-apps-image-enum

Removed icon validation in case tiles

MartinRiese committed:
Fix format_value for popup format

MartinRiese committed:
Extract constant

MartinRiese committed:
Add type hints

Jtang-1 committed:
refactor: move styling

Jtang-1 committed:
feat: align header and rows left edge

orangejenny committed:
Extracted CSS class

orangejenny committed:
Fixed js error when there's no template

MartinRiese committed:
Revert "Remove feature flag from Address Popup"

This reverts commit 36e960aacdf99a3feff8aeef024539d80ee75b79.

MartinRiese committed:
Add new feature flag for showing maps in case list

* Use FF for showing "Address Popup" format in case list config
* Use FF for showing map in case list

Jtang-1 committed:
Merge branch 'master' into sk/case-tile-group-v0

orangejenny committed:
Updated to send all template config together rather than as separate vars

orangejenny committed:
Replaced preview string with grid object

orangejenny committed:
Added grid values for remaining templates

nospame committed:
Merge pull request #33061 from dimagi/em/sscs-javascript-testing

Split screen case search: javascript testing

minhaminha committed:
Update translations

MartinRiese committed:
Update corehq/toggles/__init__.py

Co-authored-by: Jenny Schweers <orange.jenny@gmail.com>

AmitPhulera committed:
Merge pull request #33165 from dimagi/ap/es/update-sync-multiplexed

Update sync multiplexed command

biyeun committed:
complete migration changes guide and specs for bootstrap 3 to 5 with template migrations

MartinRiese committed:
Merge branch 'master' into riese/map_ff

Jtang-1 committed:
lint

MartinRiese committed:
Merge branch 'master' into riese/validate_address_popup_unique

Jtang-1 committed:
refactor: move comments to avoid repetition

Jtang-1 committed:
refactor test: uses value directly instead of assigning to variable

Jtang-1 committed:
Merge pull request #33106 from dimagi/sk/case-tile-group-v0

[USH-3286] case tile group initial

Jtang-1 committed:
refactor: clarify comments

MartinRiese committed:
List fields instead of tile fields

nospame committed:
Linter: line lengths/breaks

nospame committed:
Merge branch 'em/case-search-tiles-action-fix' of github.com:dimagi/commcare-hq into em/case-search-tiles-action-fix

biyeun committed:
add initial support for flagging javascript plugin changes in bootstrap migration

biyeun committed:
add test for flag_changed_javascript_plugins

biyeun committed:
fix other bootstrap5 tests due to changes

biyeun committed:
updates to js helpers

nospame committed:
Merge pull request #33160 from dimagi/em/case-search-tiles-action-fix

Make case tiles work with split screen case search

esoergel committed:
Update CONTRIBUTING.rst

Jtang-1 committed:
Merge branch 'master' into jt/case_tile_grouping_v2

esoergel committed:
Merge pull request #33175 from dimagi/esoergel-patch-1

Update CONTRIBUTING.rst

Jtang-1 committed:
Merge pull request #33157 from dimagi/jt/case_tile_grouping_v2

Case Tile Grouping UI

Jtang-1 committed:
feat: handles scenarios where multiple of same xpath_func exists

mjriley committed:
Fixed errors with test_view_helpers tests

Jtang-1 committed:
refactor: pull out sort element population for case tiles

MartinRiese committed:
Merge pull request #33171 from dimagi/riese/map_ff

Riese/map ff

Jtang-1 committed:
Merge branch 'master' into jt/case_tiles_sorting

Jtang-1 committed:
test: refactor and add map-popup field

Jtang-1 committed:
lint

jingcheng16 committed:
Merge pull request #33173 from dimagi/create-pull-request/update-translations-38a636c

Update Translations

Jtang-1 committed:
Merge pull request #33125 from dimagi/jt/case_tiles_sorting

Case Tiles Sorting

MartinRiese committed:
Change the max initial zoom level for case list map

MartinRiese committed:
Merge pull request #33176 from dimagi/riese/zoom_level

Change the max initial zoom level for case list map

MartinRiese committed:
Merge pull request #33169 from dimagi/riese/validate_address_popup_unique

Add validator for multiple uses of Address format

Jtang-1 committed:
Merge branch 'master' into jt/login_logout_API_params_separation

orangejenny committed:
Added defaults to CaseTileTemplateConfig

orangejenny committed:
Removed unnecessary label text

orangejenny committed:
Fixed grid-template-columns

orangejenny committed:
Merge branch 'master' into jls/case-tile-template-preview

mjriley committed:
Merge branch 'mjr/erm-update-rules' into mjr/erm-fixtures

orangejenny committed:
Fixed lint errors

shubham1g5 committed:
abstract endpoint validation in a separate method

mjriley committed:
Fixed failing ERM view helper tests

mjriley committed:
Fixed failing Lookup Table test

zandre-eng committed:
check for both case_id and owner_id

zandre-eng committed:
remove unused import

zandre-eng committed:
remove index owner check

zandre-eng committed:
unit test

AmitPhulera committed:
add missing docstring for batch size

AmitPhulera committed:
add support for purging _id in source if they are present

this codeblock has been shamlessly copied from https://github.com/dimagi/commcare-hq/pull/32588

AmitPhulera committed:
add purge_ids support in sync multiplexed

AmitPhulera committed:
fix typo

AmitPhulera committed:
switch back to original release of flower

biyeun committed:
Merge pull request #33150 from dimagi/bmb/b5-migration-script

[B5 Migration] Create initial migration tool for migrating templates from B3>B5 in an application

esoergel committed:
pr feedback

stephherbers committed:
when vlaidating fields, check the queryResponse if avaible for SSCS

stephherbers committed:
opt for cloudcare/js/formplayer/menus/collections rather than Backbone.Collection

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into es/download-file

esoergel committed:
lint

AmitPhulera committed:
Merge pull request #33179 from dimagi/ap/update-flower

Switch back to original release of flower

shubham1g5 committed:
Adds a UI field for session endpoint for case list without selection

Jtang-1 committed:
refactor: moves processing parameters into dataclass

Jtang-1 committed:
refactor: avoids converting to dict

esoergel committed:
Merge pull request #33162 from dimagi/es/download-file

Make downloadable links to server files

Jtang-1 committed:
refactor: saves qualifier as key to avoid extra splitting steps

shubham1g5 committed:
Pass in endpoint id as an argument

shubham1g5 committed:
Adds tests for verifying existing case list session endpoints

millerdev committed:
Coerce QuerySet to list

Prevents ProgrammingError: relation "motech_connectionsettings" does not exist

because ConnectionSettings and Repeater may not be in the same database.

stephherbers committed:
Merge pull request #33180 from dimagi/smh/sscs-field-validation-fix

Split Screen Case Search: field validation fix

orangejenny committed:
Applied line break fix to mapping popup

shubham1g5 committed:
Tests for new case list session endpoints

shubham1g5 committed:
Skip last selection datum for case list session endpoints

Jtang-1 committed:
lint

shubham1g5 committed:
Fix test by adding the new module property

Jtang-1 committed:
Merge pull request #33077 from dimagi/jt/login_logout_API_params_separation

Jt/login logout api params separation

shubham1g5 committed:
Retain default params to fix a test

orangejenny committed:
Merge pull request #33182 from dimagi/jls/render-line-breaks-map-popup

Applied line break fix to mapping popup

AmitPhulera committed:
Merge pull request #33178 from dimagi/ap/es/add-support-for-purging-ids

Add support for purging _id in source in ES docs

kaapstorm committed:
Unused import, missing translation

kaapstorm committed:
Lint

kaapstorm committed:
Interesting that jQuery figured this out.

kaapstorm committed:
Merge pull request #33185 from dimagi/nh/tweaks

A few tweaks to Attendance Tracking

orangejenny committed:
Fixed bad merge

orangejenny committed:
Updated docs

orangejenny committed:
Added alignment to json files and fixed person_simple's grid

orangejenny committed:
Updated js to use alignment attributes

AmitPhulera committed:
initialize ddtrace in settings

AmitPhulera committed:
merge commit

AddisonDunn committed:
replace HTML newline earlier on in logic

nospame committed:
Check if items.features is empty to prevent error on empty array

AddisonDunn committed:
turn on breaks setting for markdown-it inside render util method

AmitPhulera committed:
delete removed function and set tracer to false on every env except staging

AmitPhulera committed:
run make requirements to fix inconsistent requirement files

mjriley committed:
address PR feedback for ERM keywords and reports

mjriley committed:
Merge branch 'mjr/erm-fixtures' into mjr/erm_reports_keywords_fixes

mjriley committed:
Fix ViewHelper tests after removing overwritable flag from view models

mjriley committed:
Fixed custom data fields test to avoid hitting elasticsearch

mjriley committed:
Fixed Fixture error message text

orangejenny committed:
Updated hq_events_spec

orangejenny committed:
Updated split_screen_case_search_spec

orangejenny committed:
Updated utils_spec

orangejenny committed:
Renamed menu_list_test for consistency with other test files

orangejenny committed:
Updated integration_spec

orangejenny committed:
Updated menu_list_spec

orangejenny committed:
Updated query_spec

orangejenny committed:
Updated user_spec

orangejenny committed:
Updated debugger_spec

orangejenny committed:
Updated middleware_spec

orangejenny committed:
Updated form_ui_spec

orangejenny committed:
Updated web_form_session_spec

orangejenny committed:
Updated utils_spec

orangejenny committed:
Updated entries_spec

orangejenny committed:
Updated integration_spec

orangejenny committed:
Updated case_list_pagination_spec

AmitPhulera committed:
comment ddtrace patching

AmitPhulera committed:
update myst-parser

zandre-eng committed:
only retrieve case_id for case updates

AmitPhulera committed:
use  instead of  for searching

AmitPhulera committed:
remove unused variables

AmitPhulera committed:
don't allow to sort on _id in ESQuery

AmitPhulera committed:
linting errors

AmitPhulera committed:
refactor error message to make it more readable

AmitPhulera committed:
fix lint failures in tests

AddisonDunn committed:
refactor configuring markdownit

mjriley committed:
Fixed translation issues with ERM error messages

orangejenny committed:
Merge pull request #33135 from dimagi/jls/case-tile-template-preview

Rough case tile template preview

MartinRiese committed:
Add wrapper around case list tile select all and style like tile

ajeety4 committed:
fcm followup:resolves linting errors and sort imports

MartinRiese committed:
Merge pull request #33193 from dimagi/riese/line_up_map

Add wrapper around case list tile select all and style like tile

nospame committed:
Merge pull request #33188 from dimagi/em/geocoder-error-message-bugfix

Show correct message in geocoder widget when no results returned

orangejenny committed:
Merge pull request #33194 from dimagi/jls/tab-cloudcare-tests

Wrap cloudcare test files in hqDefine

shubham1g5 committed:
rename var

nospame committed:
Initialize caseTileTemplate to null if falsy

shubham1g5 committed:
lint

shubham1g5 committed:
Merge branch 'master' into endpointCaseList

shubham1g5 committed:
adds back import as it caused MultiSelectChildModuleDatumIDTests to fail

esoergel committed:
remove star import

esoergel committed:
flake8

esoergel committed:
isort

esoergel committed:
import from consts file

nospame committed:
Use forceManualSearch for requests from split screen case search

Commit:
Merge remote-tracking branch 'origin/master' into akj/ush-3280-DD-properties-ordering

zandre-eng committed:
Merge pull request #33177 from dimagi/ze/owner_id_case_api_permissions

Check for both case_id and owner_id in Case API

AmitPhulera committed:
suggestion from review

kaapstorm committed:
import from consts file

kaapstorm committed:
flake8

kaapstorm committed:
isort

stephherbers committed:
change button text from 'submit' to 'search' when sidebar is enabled

AddisonDunn committed:
Merge pull request #33187 from dimagi/ad/swtich-newline-syntax-web-apps

Replace HTML newline earlier on in logic

nospame committed:
Only set forceManualSearch if explicitly true

minhaminha committed:
Update translations

AmitPhulera committed:
Update corehq/apps/es/es_query.py

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera committed:
Merge pull request #33186 from dimagi/ap/update-ddtrace

Update ddtrace

AmitPhulera committed:
Merge pull request #33191 from dimagi/ap/update-myst-parser

update myst-parser

shubham1g5 committed:
merge template for case_list_session_endpoint in session_endpoint template itself

millerdev committed:
Merge pull request #33152 from dimagi/super-linter/disable-errors

Disable errors so build doesn't fail

shubham1g5 committed:
Merge pull request #33183 from dimagi/endpointCaseList

Case List Session Endpoint

kaapstorm committed:
Move comment outside string

Co-authored-by: Ethan Soergel <esoergel@users.noreply.github.com>

kaapstorm committed:
Merge pull request #33201 from dimagi/nh/lint_entries

Lint `suite_xml/sections/entries.py`

millerdev committed:
Revert "Disable errors so build doesn't fail"

Commit:
Merge remote-tracking branch 'origin/master' into akj/ush-3280-DD-properties-ordering

stephherbers committed:
Merge pull request #33202 from dimagi/smh/sscs-button-semantic-change

Split Screen Case Search: Change default button text from 'submit' to 'search' when sidebar is enabled

esoergel committed:
Merge pull request #33198 from dimagi/es/lint-models

Lint app manager models

Commit:
Merge pull request #33098 from dimagi/akj/ush-3280-DD-properties-ordering

[USH-3280] Retain order of DD group properties

AmitPhulera committed:
Merge pull request #33192 from dimagi/ap/es/fix-sort-param

Fix sort param in case v0.6

nospame committed:
Use sidebarEnabled directly instead of creating new property

nospame committed:
Merge pull request #33197 from dimagi/em/case-tile-template-quick-bugfix

Initialize caseTileTemplate to null if falsy

mjriley committed:
Merge pull request #33204 from dimagi/create-pull-request/update-translations-6608166

kaapstorm committed:
Test empty case type in Case API v0.6

nospame committed:
Merge pull request #33199 from dimagi/em/sscs-force-manual-search

Use forceManualSearch for requests from split screen case search

pxwxnvermx committed:
Forward migration for removing limits from dd description fields

millerdev committed:
Merge pull request #33210 from dimagi/dm/list-repeaters

Coerce QuerySet to list

esoergel committed:
Merge pull request #33212 from dimagi/pkv/dd-description-limit-revert

Forward migration for removing limits from dd description fields

dependabot[bot] committed:
Bump django from 3.2.19 to 3.2.20 in /requirements

Bumps [django](https://github.com/django/django) from 3.2.19 to 3.2.20.
- [Commits](https://github.com/django/django/compare/3.2.19...3.2.20)

---
updated-dependencies:
- dependency-name: django
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

millerdev committed:
Remove obsolete TestCase.icds_dbs

ajeety4 committed:
Merge pull request #33196 from dimagi/ay/push-notification-linting

resolves linting errors and sorts imports (FCM followup)

kaapstorm committed:
Revert "Test empty case type in Case API v0.6"

This reverts commit c35fd779cf8cd9a6dd2ede04e61ce4645ddbac8b.

kaapstorm committed:
Rather add test to existing class

millerdev committed:
Merge pull request #33039 from dimagi/gh/data-deletion/form-deletion-ui

Support deleting individual forms

millerdev committed:
Merge pull request #33214 from dimagi/dm/rm-icds-dbs

Remove obsolete TestCase.icds_dbs

millerdev committed:
Merge pull request #33209 from dimagi/revert-33152-super-linter/disable-errors

Revert "Disable errors so build doesn't fail"

biyeun committed:
Merge pull request #33213 from dimagi/dependabot/pip/requirements/django-3.2.20

Bump django from 3.2.19 to 3.2.20 in /requirements

kaapstorm committed:
Merge pull request #33211 from dimagi/nh/api_test

Test empty case type in Case API v0.6

mjriley committed:
Merge branch 'master' into mjr/erm-custom-roles

mjriley committed:
Created visual indicators for ERM linked applications

mjriley committed:
Only show ERM sync indicator on active push type

mjriley committed:
Fixed a javascript typo for ERM/MRM domains coexisting

mjriley committed:
Added hack to display ERM appcues in production

mjriley committed:
Added clarification comment on ERM appcues hack

mjriley committed:
Merge pull request #33154 from dimagi/mjr/erm_reports_keywords_fixes

mjriley committed:
Merge pull request #33114 from dimagi/mjr/erm-fixtures

mjriley committed:
Merge pull request #33095 from dimagi/mjr/erm-update-rules
